stan/math/rev/fun/mdivide_left.hpp

```diff
@@ -12,238 +12,48 @@
+  check_multiplicable("mdivide_left", "A", A, "B", B);
   if (A.size() == 0) {
-    return {0, b.cols()};
+    return {0, B.cols()};
   }
 
-  // NOTE: this is not a memory leak, this vari is used in the
-  // expression graph to evaluate the adjoint, but is not needed
-  // for the returned matrix.  Memory will be cleaned up with the
-  // arena allocator.
-  auto *baseVari = new internal::mdivide_left_vv_vari<
-      T1::RowsAtCompileTime, T1::ColsAtCompileTime, T2::RowsAtCompileTime,
-      T2::ColsAtCompileTime>(A, b);
+  using A_ref_t = ref_type_t<T1>;
+  using B_ref_t = ref_type_t<T2>;
 
-  Eigen::Matrix<var, T1::RowsAtCompileTime, T2::ColsAtCompileTime> res(
-      b.rows(), b.cols());
-  res.vi() = Eigen::Map<matrix_vi>(baseVari->variRefC_, res.rows(), res.cols());
+  A_ref_t A_ref = A;
+  B_ref_t B_ref = B;
 
-  return res;
-}
+  arena_matrix<promote_scalar_t<var, T1>> arena_A;
+  arena_matrix<promote_scalar_t<var, T2>> arena_B;
 
-template <typename T1, typename T2, require_eigen_vt<is_var, T1> * = nullptr,
-          require_eigen_vt<std::is_arithmetic, T2> * = nullptr>
-inline Eigen::Matrix<var, T1::RowsAtCompileTime, T2::ColsAtCompileTime>
-mdivide_left(const T1 &A, const T2 &b) {
-  check_square("mdivide_left", "A", A);
-  check_multiplicable("mdivide_left", "A", A, "b", b);
-  if (A.size() == 0) {
-    return {0, b.cols()};
+  if (!is_constant<T1>::value) {
+    arena_A = A_ref;
   }
 
-  // NOTE: this is not a memory leak, this vari is used in the
-  // expression graph to evaluate the adjoint, but is not needed
-  // for the returned matrix.  Memory will be cleaned up with the
-  // arena allocator.
-  auto *baseVari = new internal::mdivide_left_vd_vari<
-      T1::RowsAtCompileTime, T1::ColsAtCompileTime, T2::RowsAtCompileTime,
-      T2::ColsAtCompileTime>(A, b);
-
-  Eigen::Matrix<var, T1::RowsAtCompileTime, T2::ColsAtCompileTime> res(
-      b.rows(), b.cols());
-  res.vi() = Eigen::Map<matrix_vi>(baseVari->variRefC_, res.rows(), res.cols());
+  if (!is_constant<T2>::value) {
+    arena_B = B_ref;
+  }
 
-  return res;
-}
+  arena_matrix<Eigen::MatrixXd> arena_A_val = value_of(A_ref);
+  arena_matrix<Eigen::Matrix<var, T1::RowsAtCompileTime, T2::ColsAtCompileTime>>
+      res = arena_A_val.colPivHouseholderQr().solve(value_of(B_ref));
 
-template <typename T1, typename T2,
-          require_eigen_vt<std::is_arithmetic, T1> * = nullptr,
-          require_eigen_vt<is_var, T2> * = nullptr>
-inline Eigen::Matrix<var, T1::RowsAtCompileTime, T2::ColsAtCompileTime>
-mdivide_left(const T1 &A, const T2 &b) {
-  check_square("mdivide_left", "A", A);
-  check_multiplicable("mdivide_left", "A", A, "b", b);
-  if (A.size() == 0) {
-    return {0, b.cols()};
-  }
+  reverse_pass_callback([arena_A, arena_B, arena_A_val, res]() mutable {
+    Eigen::Matrix<double, T1::RowsAtCompileTime, T2::ColsAtCompileTime> adjB
+        = arena_A_val.transpose().colPivHouseholderQr().solve(res.adj());
```

Comment:
Is there a way to store things like the output of a `covPivHouseholderQr` in the arena (or the output of llt, ldlt, etc.)? In this function we recompute the decomposition in the callback cause it's not easy to save stuff.

Reply:
No idt so, looking at the code for [`ColPivHouseholderQR`](https://eigen.tuxfamily.org/dox/ColPivHouseholderQR_8h_source.html) it won't work with `Map`. But we could write a more generic `arena_object<>` that would then inherit from `chainable_alloc` and just hold an object so it's delete get's called.

Reply:
> But we could write a more generic arena_object<> that would then inherit from chainable_alloc

Hmm that does sound interesting.

Could we implemented something like a `shared_ptr` for our arena? It would track things allocated on the heap that will be destructed with `recover_memory`/`recover_memory_nested`. The shared pointers themselves would be lightweight so they could be copied around easily.

So the code here would become:

```cpp
auto hhqr = stan::math::make_shared(arena_A_val.transpose().colPivHouseholderQr());
arena_matrix<Eigen::Matrix<var, T1::RowsAtCompileTime, T2::ColsAtCompileTime>>
      res = hhqr->solve(value_of(B_ref));

reverse_pass_callback([arena_A, arena_B, arena_A_val, res, hhqr]() mutable {
    Eigen::Matrix<double, T1::RowsAtCompileTime, T2::ColsAtCompileTime> adjB
        = hhqr->solve(res.adj());
  ...
```

Reply:
You may want to check out how I implemented arena version of `matrix_cl`.